test/test_bson_corpus.py

```diff
@@ -203,6 +205,13 @@ def run_test(self):
                                              'case: ' + description)
                     except (ValueError, KeyError, TypeError, InvalidId):
                         pass
+            elif bson_type == '0x05':
+                try:
+                    decode_extjson(parse_error_case['string'])
+                    raise AssertionError('exception not raised for test '
+                                         'case: ' + description)
+                except (AttributeError, ValueError):
```

Comment:
The `AttributeError` is raised when the `$uuid` field contains a value of the incorrect type - in this case a document. 

```
ERROR: test_binary (test.test_bson_corpus.TestBSONCorpus)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/pmital/Developer/mongo-python-driver/test/test_bson_corpus.py", line 210, in run_test
    decode_extjson(parse_error_case['string'])
  File "/Users/pmital/Developer/mongo-python-driver/bson/json_util.py", line 409, in loads
    return json.loads(s, *args, **kwargs)
  File "/Users/pmital/.pyenv/versions/3.7.5/lib/python3.7/json/__init__.py", line 361, in loads
    return cls(**kw).decode(s)
  File "/Users/pmital/.pyenv/versions/3.7.5/lib/python3.7/json/decoder.py", line 337, in decode
    obj, end = self.raw_decode(s, idx=_w(s, 0).end())
  File "/Users/pmital/.pyenv/versions/3.7.5/lib/python3.7/json/decoder.py", line 353, in raw_decode
    obj, end = self.scan_once(s, idx)
  File "/Users/pmital/Developer/mongo-python-driver/bson/json_util.py", line 408, in <lambda>
    pairs, json_options)
  File "/Users/pmital/Developer/mongo-python-driver/bson/json_util.py", line 428, in object_pairs_hook
    return object_hook(json_options.document_class(pairs), json_options)
  File "/Users/pmital/Developer/mongo-python-driver/bson/json_util.py", line 452, in object_hook
    return _parse_legacy_uuid(dct, json_options)
  File "/Users/pmital/Developer/mongo-python-driver/bson/json_util.py", line 494, in _parse_legacy_uuid
    return uuid.UUID(doc["$uuid"])
  File "/Users/pmital/.pyenv/versions/3.7.5/lib/python3.7/uuid.py", line 157, in __init__
    hex = hex.replace('urn:', '').replace('uuid:', '')
AttributeError: 'SON' object has no attribute 'replace'
```

Since this is a fairly hard to understand error message for a user, I think we should manually validate that the type is correct (and raise something more sensible like a TypeError). Would you agree?

Reply:
Agreed.